src/Parser.hs

```diff
@@ -0,0 +1,299 @@
+import Data.Data
+import Data.Complex
+
+
+
+data Token  = Lambda | If | Set | POpen | PClose | Identifier String | Quote | ShortQuote | Integral Int | Rational Int Int
+            | Real Double | String String | Complex (Complex Double)  | Bool Bool | Char Char
+            | Dot | QuasiQuote | ShortQuasiQuote | Unquote | ShortUnquote | UnquoteSplice | CommentDatum
+            | ShortUnquoteSplice | Label Integer | LabelRef Integer | Define deriving (Eq, Show) 
+
+data MetaNode = LambdaNode [MetaNode] MetaNode [MetaNode] | PairNode MetaNode MetaNode 
+            | RealAtom Double | IntegralAtom Int | RationalAtom Int Int | EmptyAtom
+            | StringAtom String | ComplexAtom (Complex Double) | BoolAtom Bool | CharAtom Char 
+            | IdentifierAtom String | ApplicationNode MetaNode [MetaNode] 
+            | IfNode MetaNode MetaNode MetaNode | SetNode MetaNode MetaNode | DefineNode MetaNode MetaNode 
+            deriving (Eq, Show)
+
+parseScheme :: [Token] -> MetaNode
+parseScheme st = case runState (parseExpression "Scheme Program")  st of
+    (mn, []) -> mn
+    (_, t:_) -> error $ "Unexpected token " ++ show t ++  " not allowed in top level of program"
+
+push :: Token -> State [Token] ()
+push t = do
+    ts <- get
+    put (t:ts)
+
+pull :: String -> State [Token] Token
+pull context = do
+    ts <- get
+    case ts of 
+        (t:ts) -> do
+            put ts
+            return t
+        _ -> error $ "Unexpected end of token stream in " ++ context
+
+pullEq :: String -> Token -> State [Token] ()
+pullEq context c = do
+    ts <- get
+    case ts of
+        (t:ts) -> do
+            put ts
+            if  t ==  c then
+                return ()
+            else
+                error $ (show c) ++ " in " ++ context ++ " expected, but not found"
+        _ -> error $ "Unexpected end of token stream in " ++ context
+
+peek :: String -> State [Token] Token
+peek context = do
+    ts <- get
+    case ts of
+        (t:ts) -> return t 
+        _ -> error $ "Unexpected end of token stream in " ++ context 
+
+parseExpression :: String -> State [Token] MetaNode
+parseExpression context = do
+    t <- peek context
+    case t of
+        POpen -> parseSyntax
+        ShortQuote -> do
+            pullEq "Datum" ShortQuote
+            parseQuotedDatum
+        ShortQuasiQuote -> do
+            pullEq "Quasiquoted Datum" ShortQuasiQuote
+            parseQuasiQuotedDatum
+        CommentDatum -> do
+            pullEq "Commented Datum" CommentDatum
+            parseExpression "Commented Datum"
+            parseExpression context
+        _ -> parseAtom
+
+parseSyntax :: State [Token] MetaNode
+parseSyntax = do
+    pullEq "Syntactic construct" POpen 
+    t <- peek "Syntactic construct"
+    case t of
+        Lambda -> parseLambda
+        Define -> parseDefine
+        Quote -> parseQuote
+        QuasiQuote -> parseQuasiQuote
+        If -> parseIf
+        Set -> parseSet
+        _ -> parseApplication
```

Comment:
can this be `Identifier -> parseApplication` followed by `_ -> error ...`?

Reply:
No the Scheme grammar does not allow this